Lib/ntpath.py

```diff
@@ -640,23 +638,68 @@ def commonpath(paths):
         genericpath._check_arg_types('commonpath', *paths)
         raise
 
+MAX_PATH = 260
 
-# determine if two files are in fact the same file
 try:
-    # GetFinalPathNameByHandle is available starting with Windows 6.0.
-    # Windows XP and non-Windows OS'es will mock _getfinalpathname.
-    if sys.getwindowsversion()[:2] >= (6, 0):
-        from nt import _getfinalpathname
-    else:
-        raise ImportError
-except (AttributeError, ImportError):
-    # On Windows XP and earlier, two files are the same if their absolute
-    # pathnames are the same.
-    # Non-Windows operating systems fake this method with an XP
-    # approximation.
-    def _getfinalpathname(f):
-        return normcase(abspath(f))
+    from nt import _getfinalpathname
+
+    # determine if two files are in fact the same file
+    def realpath(filename):
+        filename = os.fspath(filename)
+        is_str = isinstance(filename, str)
+        if not is_str:
+            filename = os.fsdecode(filename)
+        extended_path_prefix = '\\\\?\\'
+        is_extended_path = filename.startswith(extended_path_prefix)
+        unresolved = filename if is_extended_path else abspath(filename)
+        resolved_parts = []
+        while True:
+            try:
+                resolved_parts.append(_getfinalpathname(unresolved))
+                break
+            except OSError:
+                unresolved, tail = split(unresolved)
+                if not tail:
+                    resolved_parts.append(unresolved)
+                    break
+                resolved_parts.append(tail)
+        resolved = join(*reversed(resolved_parts))
+        # try to convert extended path to normal if
+        # initial path did not use \\?\ prefix and result uses it
+        if not is_extended_path and resolved.startswith(extended_path_prefix):
+            resolved = _extended_to_normal(resolved)
+        return resolved if is_str else os.fsencode(resolved)
+
+    def _extended_to_normal(path):
+        letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+        if (len(path) > 7 and
+            path[4].upper() in letters and
+            path[5] == ':' and
+            path[6] == '\\'):
+            # extended path with \\?\ prefix
+            # 4 is len('\\?\')
+            normal_path = path[4:]
+        elif len(path) > 8 and path[:8].upper() == '\\\\?\\UNC\\':
+            # UNC path with \\?\ prefix - drop prefix
+            # 7 is len('\\?\UNC')
+            normal_path = '\\' + path[7:]
+        else:
+            # not a UNC or drive-letter path
+            # return path as-is
+            return path
 
+        if len(normal_path) < MAX_PATH and normal_path == abspath(normal_path):
+            return normal_path
```

Comment:
`realpath` shouldn't return a normal path with a trailing backslash, but `_getfinalpathname` leaves a trailing backslash on a UNC path (well, the SMB redirector does when we query `FileNameInformation`).  We can end up with this case if a mapped drive or symlink resolves to a UNC path. We should remove any trailing backslash here if it's not a drive root directory. The most direct approach would be to call `normpath`. In that case I suggest verifying that it's a properly normal path via `_getfullpathname` instead of `abspath`. That splits up the cost of the replaced `abspath` call.

Reply:
if I understand you correctly it should become:
```python
if len(normal_path) < MAX_PATH and normal_path == _getfullpathname (normal_path):
    return normpath(normal_path)
```
?

Reply:
Yes, and also the assignments become, for example, `normal_path = normpath(path[4:])`. Unless you want to inline the work of `normpath`.